xdsl/ir/affine/affine_expr.py

```diff
@@ -180,6 +180,13 @@ def _simplify_mul(self, other: AffineExpr) -> AffineExpr | None:
         if isinstance(self, AffineBinaryOpExpr) and self.kind == AffineBinaryOpKind.Mul:
             if fold := self.rhs._try_fold_constant(other, AffineBinaryOpKind.Mul):
                 return self.lhs * fold
+        # Fold (expr + expr) * constant.
+        if (
+            isinstance(self, AffineBinaryOpExpr)
+            and self.kind == AffineBinaryOpKind.Add
+            and isinstance(other, AffineConstantExpr)
+        ):
+            return self.lhs * other + self.rhs * other
```

Comment:
This little addition makes the generated affine maps much more readable:
changes
`((((d0 * 2) + d1) * 2 + d2) * 2 + d3) * 2)`
into
`(d0 * 16) + (d1 * 8) + (d2 * 4) + (d3 * 2)`

I don't know if this is legal or if this needs to be aligned with MLIR in some way?

Reply:
Do you know if MLIR does this? I personally am not a fan of the automatic optimisation, but this one seems safe enough

Reply:
> I personally am not a fan of the automatic optimisation

Isn't this ... what a compiler does?

There are some automizations in `simplify_mul` here: https://github.com/llvm/llvm-project/blob/e67f2cc3fc38cec2041cfb197ac4688ed3d16e7e/mlir/lib/IR/AffineExpr.cpp#L769, but it doesn't seem to implement this exact one. But maybe it should?

I just did this because I know running `simplify` in matlab (no judging) does exactly this, and is really nice.

Reply:
My stance is it would be nice for this to be explicit, and I'd be happier if there were a simplify method like in matlab :) I'm happy to add this one since the others are already there

Reply:
I think its also reasonable (if easy to do) that an affine map expressing something should only have one single canonical representation